shuffle index list with numpy, scatter list, use file for large lists

adammoody opened the pull request:
This updates the index selection logic in a couple ways:
- switches from ``random.shuffle`` to ``numpy.random.shuffle`` (timed to be about 10x faster)
- replaces the broadcast with a scatter

The scatter is implemented in two ways.  The first uses a scatter collective.  The second writes the full list to a temporary file that is then read back by each process.

I suppose if the scatter collective seems to be reliable, the temporary file is not needed and could be dropped.  I've left it in there for now.

Update:
Decided to drop the scatter file to simplify the code.  I think the scatterv should be sufficient, and if not, it's easy to add the file method back.

adammoody commented:
@thomasw21 , and would you please also take a look at this one if you have some time?  It's on the shorter side.  Thanks!

BTW, I've got a commit ready to push that drops the scatter-via-file implementation.  I'm leaning toward dropping that to simplify the code.  I think the ``scatterv_`` should be sufficient.  I haven't seen a problem yet in my testing on ``oscar``.  If we do hit a problem with scatter, the file-based method is easy to add back.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +219,21 @@ def barrier(args):
-
-        # allocate a tensor of appropriate size
-        # initialize tensor with list values on root
+        tensors = []
```
```suggestion
        chunks = []
```

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +219,21 @@ def barrier(args):
-        # broadcast tensor from root, and return as a new list
-        dist.broadcast(tvals, src=root)
-        return tvals.tolist()
+        tensor = torch.from_numpy(outval)
```
~Isn't there an issue where the size of tensors don't match anymore?~ Could you add a check that outval, correspond to the same size as corresponding count?

adammoody replied:
Yes, that name is not very descriptive rename.  And good idea to check the size of the output array.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -330,63 +325,75 @@ def load_dset(args):
-    remainder = num % num_ranks
-    if rank < remainder:
-        start = (num_per_rank + 1) * rank;
-        end = start + (num_per_rank + 1)
+            idxlist = idxlist[:args.count]
```
```python
idxlist = idxlist[:num_samples]
```

adammoody replied:
Got it.  Thanks.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -330,63 +325,75 @@ def load_dset(args):
+    idx = np.zeros(counts[args.rank], np.int64)
+
+    # scatter index list if small enough
+    scatter_limit = 20000000
```
Why do we need this? Also if you don't mind can you use `20_000_000` it's easier to read IMO.

adammoody replied:
This will go away.  That 20 million was an arbitrary number I picked out of the air anyway.  Thanks for the tip on the underscores for readability.  I'll do that if this has to come back.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -330,63 +325,75 @@ def load_dset(args):
-        start = (num_per_rank + 1) * remainder + num_per_rank * (rank - remainder);
-        end = start + num_per_rank
-    return start, end
+        # The index list is too big to send to every process.
```
Is there such a thing? That we can't load all indices in memory? 8bytes per int64, so `20_000_000` elements represent around 160Mb which should be more than okay.

adammoody replied:
I'm more worried about the communication method being used.  We can revisit this if it happens to show up.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -330,63 +325,75 @@ def load_dset(args):
+        idx_count = counts[args.rank]
+        if idx_count > 0:
+            with open(indexlistfile, "rb") as f:
+                idx_start = sum(counts[:args.rank])
```
We should compute those offsets only once and dispatch to every rank.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -330,63 +325,75 @@ def load_dset(args):
+        if idx_count > 0:
+            with open(indexlistfile, "rb") as f:
+                idx_start = sum(counts[:args.rank])
+                offset = idx_start * 8
```
Add a comment that this is linked to int64.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -330,63 +325,75 @@ def load_dset(args):
+        if args.rank == 0:
+            os.remove(indexlistfile)
+
+        barrier(args)
```
I don't think this last barrier is needed, you could just let other ranks go to the next step without worrying no?

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -424,15 +431,13 @@ def rank_files_write(args, dset, idx, encoder):
             for key in args.columns:
                 # tokenize text for the given sample index
-                text = dset[i][key]
+                text = dset[sample_id][key]
```
I think you can even run `dset[idx]` if I'm not wrong, to be confirmed. If so I'd suggest something like

```python
for row in dset[idx]:
     for key in args.columns:
          text = rows[key]
```

adammoody replied:
Oh, that's useful to know.  If you don't mind, I'd like to stick with the harder way for now, because I'm using this same code to support a JSON file and that particular code only supports an ``int`` for the argument in the ``__get__`` call.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -451,11 +456,11 @@ def rank_files_write(args, dset, idx, encoder):
                 timestamp = time.strftime("%Y-%m-%dT%H:%M:%S")
                 docs = dset_stats[0] * args.numranks
-                percent = docs / len(idx) * 100.0
+                percent = docs / num_samples * 100.0
```
Isn't that just `len(dset)`? Unless you want to take in account count also. I don't think it's worth adding a parameter to the signature of the method, since you can recover it from `args.count` and `len(dset)`.

Seeing as it's already an approximation, you could say that `percent = dset_stats[0] / len(idx) * 100`

adammoody replied:
Right, this is to adjust for count being less than ``len(dset)``.  I'll create a separate function to compute ``num_samples`` and drop it from the param list.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -581,29 +586,29 @@ def main():
         print(dset)
         print("Selecting features:", args.columns)
 
+    args.level = "document"
+    if args.split_sentences:
+        args.level = "sentence"
```
Let's move this in get_args

adammoody replied:
Yep, done.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -451,11 +456,11 @@ def rank_files_write(args, dset, idx, encoder):
                 mbs = dset_stats[2] * args.numranks / elapsed / 1024 / 1024 if elapsed > 0.0 else 0.0
-                secs_left = int((len(idx) - docs) / docrate if docrate > 0.0 else 0.0)
-                print(f"{timestamp}: Processed (estimated) {docs} of {len(idx)} docs ({percent:0.2f}%),",
+                secs_left = int((num_samples - docs) / docrate if docrate > 0.0 else 0.0)
```
ditto

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -451,11 +456,11 @@ def rank_files_write(args, dset, idx, encoder):
-                secs_left = int((len(idx) - docs) / docrate if docrate > 0.0 else 0.0)
-                print(f"{timestamp}: Processed (estimated) {docs} of {len(idx)} docs ({percent:0.2f}%),",
+                secs_left = int((num_samples - docs) / docrate if docrate > 0.0 else 0.0)
+                print(f"{timestamp}: Processed (estimated) {docs} of {num_samples} docs ({percent:0.2f}%),",
```
you could extrapolate also? Overall I feel it's a shame to add arguments for the sake of a logger. If you really want to keep it, I'm also okay with that.

thomasw21 reviewed (commented):
Interesting, I'm okay with this PR, but I'd really consider looking into HF shuffle. Despite being slow (and we're not even sure if we're much slower than here), it requires a lot less code IMO.

adammoody commented:
> 
> 
> Interesting, I'm okay with this PR, but I'd really consider looking into HF shuffle. Despite being slow (and we're not even sure if we're much slower than here), it requires a lot less code IMO.

Yes, it's still worth looking at the HF dataset shuffle.  I can still look at that in the future.  However, having some native shuffle support is useful, because PR https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/60 adds support for JSON files.  If we find that the HF dataset shuffle is better, I can use that with a HF dataset and fallback to this for a JSON file.

I'll also go ahead and push the change to drop the scatter file, which simplifies the code quite a bit.

adammoody commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +219,21 @@ def barrier(args):
-            tvals = torch.tensor(vals, dtype=torch.int64)
-        else:
-            tvals = torch.zeros(length[0], dtype=torch.int64)
+            for rank in range(args.numranks):
+                start = sum(counts[:rank])
+                end = start + counts[rank]
+                tensors.append(torch.tensor(invals[start:end]))
```
Cool.  Thanks for the tip on ``torch.split``.  That looks to be perfect.

adammoody commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +219,21 @@ def barrier(args):
-    """Broadcast list of vals from root to all ranks, returns newly allocated list"""
+def scatterv_(args, invals, counts, outval, root=0):
+    """Scatter values from invals according to counts array, receive values in outval"""
     if args.use_mpi:
```
Yes, thanks.

thomasw21 commented:
You can actually load json files using HF datasets. https://huggingface.co/docs/datasets/loading_datasets.html#json-files

adammoody commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +219,21 @@ def barrier(args):
     if args.use_mpi:
-        vals = args.mpi_comm.bcast(vals, root=root)
-        return vals
+        displ = [sum(counts[:rank]) for rank in range(args.numranks)]
```
Thanks.  I made that change.

adammoody commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +219,21 @@ def barrier(args):
-def bcast(args, vals, root=0):
-    """Broadcast list of vals from root to all ranks, returns newly allocated list"""
+def scatterv_(args, invals, counts, outval, root=0):
+    """Scatter values from invals according to counts array, receive values in outval"""
```
Added a note to the docstring about that.

adammoody commented:
> 
> 
> You can actually load json files using HF datasets. https://huggingface.co/docs/datasets/loading_datasets.html#json-files

I did give that a try, even by accident a few times.  For the oscar json, the load time is very slow.  I haven't investigated the reason yet.  I can imagine it may be scanning the json file to identify line breaks, meaning it's probably reading the full file from every rank.

One of the additions in PR https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/60 is that it implements a parallel scan of the json file, so the full file can be scanned and indexed in just a few seconds.  The ranks scan the file collectively rather than having each rank read the full file.

adammoody commented:
@thomasw21 , when you get a chance, would you please take another look at this one?  Some of your suggestions became moot since I dropped the scatter file, but I think I've addressed the rest.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +223,21 @@ def barrier(args):
+def scatterv_(args, invals, counts, outval, root=0):
+    """Scatter int64 values from invals according to counts array, receive values in outval"""
+    assert len(counts) == args.numranks, f"Length of counts list {len(counts)} does not match number of ranks {args.numranks}"
+    assert outval.shape == (counts[args.rank],), f"Rank {args.rank}: output buffer is of shape {outval.shape}, expected {(counts[args.rank],)}"
```
Just to make sure, if this raises an exception, other ranks will keep going and deadlock? Or do we need the same pattern as everyone where else?

adammoody replied:
Ah, yes.  Good point. That would deadlock if one rank failed the assert and the others did not.

Let's leave it this way for now in this PR.  The ``preprocess_dataset_mpi.py`` script is the only one calling ``scatterv_`` right now, and it should be fine.  Longer term, I've got an idea for the ``DistData`` class to help with these collective assert checks.

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +223,21 @@ def barrier(args):
-        # broadcast tensor from root, and return as a new list
-        dist.broadcast(tvals, src=root)
-        return tvals.tolist()
+            scatterlist = list(torch.split(torch.from_numpy(invals), counts))
```
So leaving the tuple didn't work here?

adammoody replied:
Without the ``list()`` in there, it throws this error:
```
RuntimeError: Invalid function argument. Expected parameter `scatter_list` to be of type List[torch.Tensor].
```

thomasw21 reviewed (approved):
LGTM overall! Two small questions before merging!

thomasw21 commented:
Btw there's this PR that might peak your interest: https://github.com/huggingface/datasets/pull/2747

thomasw21 commented on tools/preprocess_dataset_mpi.py:
```diff
@@ -219,26 +223,21 @@ def barrier(args):
-
-        # allocate a tensor of appropriate size
-        # initialize tensor with list values on root
+        scatterlist = []
```
```suggestion
    scatterlist = None
```

I think it's better to use None here as it's a value that should only be defined in root.

adammoody replied:
Thanks.  Ran a quick test to also verify, so I think we're good to merge.

thomasw21 reviewed (commented):
I can merge as soon as you give me the go

adammoody commented:
> 
> 
> Btw there's this PR that might peak your interest: [huggingface/datasets#2747](https://github.com/huggingface/datasets/pull/2747)

Oh, nice.  Thanks for the pointer.
